packages/next/next-server/lib/router/router.ts

```diff
@@ -491,18 +493,30 @@ export default class Router implements BaseRouter {
     if (!(options as any)._h && this.onlyAHashChange(cleanedAs)) {
       this.asPath = cleanedAs
       Router.events.emit('hashChangeStart', as)
+      // TODO: do we need the resolved href when only a hash change?
       this.changeState(method, url, as, options)
       this.scrollToHash(cleanedAs)
       this.notify(this.components[this.route])
       Router.events.emit('hashChangeComplete', as)
       return true
     }
 
-    const parsed = tryParseRelativeUrl(url)
+    const pages = await this.pageLoader.getPageList()
```

Comment:
Can we shortcut this code path if the page and url are an exact match already? We don't allow rewriting exact matches IIRC.

Reply:
We can't know if the URL is an exact match for a page until we load the page list/buildManifest so I don't believe this path can be shortcut, loading the rewrites was being shortcut if there was an exact match although it was recommended to store the client rewrites in the buildManifest so this shortcutting was removed